Add an exam checklist document (#276)
UCL/rsd-engineeringcourse: exam-checklist/knowledge-checklist.tex

\section{Knowledge Checklist}

Topics are divided up as best we can, to group things together logically.

Note that throughout, there is a general standing rule of ``you should know (abstract concept)'' also meaning ``you should be able to apply / recognise (abstract concept)'' if you're given a practical example!

\subsection{Fundamentals of Version Control}

You should understand the importance, potential advantages, and fundamental concepts of using version control for your (software) projects.

You should be familiar with fundamental \codeWord{git} functionality, workflow, and terminology.

\begin{itemize}
  \item What a repository is.
  \item The staging area, and the \codeWord{add} command.

>   \item The repository history, what "a commit" is, the \codeWord{commit} command, and how to review history either graphically or via the \codeWord{log} command.

Reviewer: ```suggestion
  \item The repository history, what ``a commit'' is, the \codeWord{commit} command, and how to review history either graphically or via the \codeWord{log} command.
```

  \item Remotes, and the \codeWord{push}, \codeWord{pull}, and \codeWord{clone} commands.
  \item Branches, and the \codeWord{checkout} / \codeWord{switch}, \codeWord{branch} commands.
  \item Merges, the \codeWord{merge} command, and merge conflicts (how they arise, what they are, how they're resolved).
\end{itemize}

\subsection{Collaborative Software Projects}

You should understand the distinction between GitHub and Git.

You should be aware of important metadata files that should be provided as part of a software project being hosted on GitHub, and the reason for including them.
In particular:

\begin{itemize}

>   \item The \codeWord{README} file, and it's inclusion of the basic ``need-to-know'' introduction that a newcomer to the project requires.

Reviewer: ```suggestion
  \item The \codeWord{README} file, and its inclusion of the basic ``need-to-know'' introduction that a newcomer to the project requires.
```

You should understand what a fork of a GitHub repository is, and how to make contributions to open-source projects using them.
You should understand why external collaborators are expected to use forks, rather than direct access via clones, to contribute to projects.

You should have a clear understanding, and practice participating in, the development life cycle.
You should be able to justify why, and explain how, each step of the life cycle is conducted.

\begin{itemize}
  \item The use of issues for tracking work to be done, reporting bugs, or having discussions about new features.
  \item The use of branches for feature development.
  \item Use of pull-requests (PRs) for suggesting changes.
  \item Importance of code review prior to approval and PR merging.
  \item Any additional safeguards; such as testing, continuous integration workflows, or documentation checking.
\end{itemize}

You should understand that there are different approaches to development; in particular the Agile and Waterfall strategies.

\subsection{Python Knowledge}

> You should have a firm grasp of the fundamentals of the Python language, including it's syntax.

Reviewer: ```suggestion
You should have a firm grasp of the fundamentals of the Python language, including its syntax.
```

You should be able to design tests based off these principles, given a particular codebase or function to work from.
You should also be able to recommend changes, additions (and possibly deletions) to test coverage / cases based on your understanding of the above.

You should be aware of the connections between a test suite and documentation, continuous integration, debugging, and PR review.
You should also be aware of how to exploit these connections to their full benefit.

You should be able use the bisection method, and \codeWord{git bisect}, to locate bugs within history.
This process may possibly be supported by an existing test suite.

You should know, understand, and be able to explain the principles of test-driven development.

\subsection{Documentation}

You should understand why it is important for all software projects to provide some form of documentation for both users and developers / maintainers.
You should also know (in the abstract) what information you expect to find / provide within project documentation.

> You should be aware that there are existing tools for generating rich-format documentation (HTML / websites etc) from plain-text files and source code.

Reviewer: ```suggestion
You should be aware that there are existing tools for generating rich-format documentation (HTML / websites, etc) from plain-text files and source code.
```

You should also be able to write docstrings in a particular format / style, given a codebase with existing examples or developer documentation that provides a clear style.

\subsection{Packaging and Libraries}

You should know what it means for a project to depend on, or be a dependency of, another library or package.

You should be able to assess both the benefits and potential drawbacks of having a project depend on an external (or 3rd-party) library.
You should also know what to look for in an external library to help you make this decision; such as the quality of their tests and documentation, how well-maintained or frequently updated it is, whether it provides too little / too much functionality outside of what you need, etc.

You should understand the difference between a collection of scripts and a package / library.
You should be able to write scripts that use functionality from a library / package that you have written or been provided.

\subsection{Code Structure and Design}

You should be able to recognise abstract design patterns in particular codebases, and be able to write code to adhere to a particular design pattern.

> You are not expected to memorise specific design patterns, but should be able to identify whether a design pattern is being used (or not) given it's description.

Reviewer: ```suggestion
You are not expected to memorise specific design patterns, but should be able to identify whether a design pattern is being used (or not) given its description.
```

You should know what the term ``refactoring'' or ``to refactor'' means in the context of a software project.
You should be able to explain why refactoring can be useful, and identify candidates for refactoring / undertake this yourself if given a particular codebase.
You should also be aware of any features that should already be in place before undertaking refactoring (for example, tests!).

\begin{figure*}
  \centering
  \includegraphics[scale=0.25]{images/beaver-refactorings.jpg}
  \caption{\label{fig:beaver-refactoring} Refactoring offers many advantages, but can be a major time-sink if the codebase has already grown quite large without any design structure being imposed at the beginning.}
\end{figure*}

\subsection{Profiling and Benchmarking}

You should be aware of the distinction between profiling and benchmarking code.

Profiling is timing the execution of a particular block of code; with the interest being in exploring how long the program spends on each particular line of code or within specific sub-functions, and how frequently these parts of the codebase are reached by the program.

> Profiling is typically used to identify "bottlenecks" within the codebase, or to get an overview of the relative importance of improving the efficiency of one function over another.

Reviewer: ```suggestion
Profiling is typically used to identify ``bottlenecks'' within the codebase, or to get an overview of the relative importance of improving the efficiency of one function over another.
```

Benchmarking is typically used to demonstrate how well a program scales onto larger data / input sizes, or provide results for comparison against projects that offer similar functionality (but perhaps in alternative programming languages).
Benchmarking can also flag when our code is performing - as a whole - above or below a certain expectation, and at which kinds of inputs that we see these deviations from theory / expectation.

You should understand \href{https://en.wikipedia.org/wiki/Big_O_notation}{``big-O'' notation} ($O(N^3)$, $O(a^N)$ etc) and its applications to benchmarking.
You should understand what it means for a quantity (say $t$) to scale exponentially or as a power-law, with another quantity (say $N$).
In each case, you should be able to infer a scaling relationship, and potentially estimate the various quantities that define the relationship, from plots or numerical information.

\begin{itemize}
  \item Power-law: $t$ scales $O(N^\alpha)$, so we expect that $t = \beta N^{\alpha}$ for some constants $\alpha, \beta$.

>         The term "polynomial growth" or "grows polynomially" is sometimes used for the case where $\alpha > 0$.

Reviewer: ```suggestion
        The term ``polynomial growth'' or ``grows polynomially'' is sometimes used for the case where $\alpha > 0$.
```

        Equivalently, we could write
        \begin{align*}
          \log t & = \alpha \log N + \log\beta
        \end{align*}
        and check if we observe a linear trend when plotting $\log t$ (y-axis) against $\log N$ (x-axis) on normal axes.
        \begin{figure*}[h!]
          \centering
          \includesvg[scale=0.45]{images/power-law-scaling.svg}
          \caption{
            \label{fig:power-law-quantity-estimation}
            Estimating the constants $\alpha ( = 1.5)$ and $\beta ( = 3)$ from a benchmarking plot (of $t = 3N^{1.5}$).
          }
        \end{figure*}
  \item Exponential: $t$ scales $O(\alpha^N)$, so we expect that $t = \beta \alpha^{N}$ for some constants $\alpha, \beta$.
        This scaling relationship looks linear on a plot of $\log t$ (y-axis) against $N$ (x-axis) when using log-log axes.
        The gradient of said linear line is an estimation for $\log \alpha$, and the y-axis intercept an estimation for $\log \beta$.
        \begin{figure*}[h!]
          \centering
          \includegraphics[scale=0.25]{images/doge-exp-vs-power.jpg}

>           \caption{\label{fig:doge-scaling} Exponential growth is always faster than polynomial growth, even if polynomial growth starts off strong.}

Reviewer: I think this caption needs more explanation. I don't really understand the image itself.

Author: I've swapped it out now for a [new plot similar to the power law scaling one](https://github.com/UCL/rsd-engineeringcourse/blob/wgraham/exam-checklist/exam-checklist/images/exp-law-scaling.svg), since if one is explained then the other should be too.
